src/D3MOracle.t.sol

```diff
@@ -0,0 +1,149 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.14;
+
+import "ds-test/test.sol";
+
+import {D3MOracle} from "./D3MOracle.sol";
+
+interface Hevm {
+    function warp(uint256) external;
+
+    function store(
+        address,
+        bytes32,
+        bytes32
+    ) external;
+
+    function load(address, bytes32) external view returns (bytes32);
+}
+
+contract D3MTestVat {
+    uint256 public live = 1;
+
+    function cage() external {
+        live = 0;
+    }
+}
+
+contract D3MTestHub {
+    uint256 c = 0;
+
+    function culled(bytes32) external view returns (uint256) {
+        return c;
+    }
+
+    function cull() external {
+        c = 1;
+    }
+
+    function uncull() external {
+        c = 0;
+    }
+}
+
+contract D3MOracleTest is DSTest {
+    Hevm hevm;
+
+    D3MTestVat vat;
+    D3MTestHub hub;
+    D3MOracle oracle;
+
+    uint256 internal constant WAD = 10 ** 18;
+
+    function setUp() public {
+        hevm = Hevm(
+            address(bytes20(uint160(uint256(keccak256("hevm cheat code")))))
+        );
+
+        vat = new D3MTestVat();
+        hub = new D3MTestHub();
+
+        oracle = new D3MOracle(address(vat), bytes32("random"));
+        oracle.file("hub", address(hub));
+    }
+
+    function test_rely_deny() public {
+        assertEq(oracle.wards(address(123)), 0);
+        oracle.rely(address(123));
+        assertEq(oracle.wards(address(123)), 1);
+        oracle.deny(address(123));
+        assertEq(oracle.wards(address(123)), 0);
+    }
+
+    function testFail_unauth_rely() public {
+        oracle.deny(address(this));
+        oracle.rely(address(123));
+    }
+
+    function testFail_unauth_deny() public {
+        oracle.rely(address(123));
+        oracle.deny(address(this));
+        oracle.deny(address(123));
+    }
+
+    function test_file_hub() public {
+        assertEq(oracle.hub(), address(hub));
+        oracle.file("hub", address(123));
+        assertEq(oracle.hub(), address(123));
+    }
+
+    function testFail_unauth_file_hub() public {
+        oracle.deny(address(this));
+        oracle.file("hub", address(123));
+    }
+
+    function testFail_vat_caged_file_hub() public {
+        vat.cage();
+        oracle.file("hub", address(123));
+    }
+
+    function test_peek() public {
+        (uint256 value, bool ok) = oracle.peek();
+        assertEq(value, WAD);
+        assertTrue(ok);
+        hub.cull();
+        (value, ok) = oracle.peek();
+        assertEq(value, WAD);
+        assertTrue(ok);
+        vat.cage();
+        (value, ok) = oracle.peek();
+        assertEq(value, WAD);
+        assertTrue(!ok);
+        hub.uncull();
+        (value, ok) = oracle.peek();
+        assertEq(value, WAD);
+        assertTrue(ok);
+    }
```

Comment:
Missing a `peek` check for when `vat` is not caged but `hub` is culled (should be !ok).

Reply:
Not sure about this one, this is exactly the case I wanted to avoid reporting `false`. That's why originally I wasn't including `vat.live` in the check and then doing `peek` behavior different than `read`.

If the system is in normal operation, and the collateral is culled we do not gain anything on start reporting an invalid price to the spotter. That will make the vat.spot for that ilk is set to 0. I don't see that can be a problem as once the `ilk` is culled in the `Hub` is irreversible for normal operation, but feeling we better keep the 1 value in vat.spot.

Reply:
Right.